### DISCUSSION TITLE
Trying to cross-compile OpenBLAS for ARM without Fortran compiler

### DISCUSSION THREAD

**Discussion by Desperado17** [OpenMathLib/OpenBLAS#4386] [Q&A]
Trying to build OpenBlas 0.3.25 with a Linaro based Aarch64 cross-compiler chain that has no Fortran compiler. CMake  gets the configure flags 

-DTARGET=ARMV8 -DBUILD_SHARED_LIBS=1

according to

https://github.com/OpenMathLib/OpenBLAS/pull/3539

Should be able to build lapack through f2c, right? Apparently, there is still a problem:

[ 82%] Built target LAPACKE
[ 82%] Linking Fortran shared library lib/libopenblas.so
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: Relocations in generic ELF (EM: 183)
/usr/bin/ld: interface/CMakeFiles/interface.dir/CMakeFiles/scopy.c.o: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/openblas_shared.dir/build.make:12709: lib/libopenblas.so.0.3] Error 1
make[1]: *** [CMakeFiles/Makefile2:428: CMakeFiles/openblas_shared.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

Am I doing something wrong?

Regards

**Comment by martin-frbg** [comment]
In theory it should work. Are you sure that it is picking up the correct toolchain on your system ? (You may have to add something like `-DCMAKE_C_COMPILER=...`)

**Reply by Desperado17** [reply]
CMAKE_C_COMPILER: aarch64-linux-gnu-gcc
CMAKE_CXX_COMPILER: aarch64-linux-gnu-g++

Looks like it.


This is how it ends when I build static:


[ 82%] Building C object utest/CMakeFiles/openblas_utest.dir/test_kernel_regress.c.o
/proprietary/dev/src/system/ext-openblas/1/workspace/OpenBLAS-0.3.25/utest/test_kernel_regress.c: In function '__ctest_kernel_regress_skx_avx_run':
/proprietary/dev/src/system/ext-openblas/1/workspace/OpenBLAS-0.3.25/utest/test_kernel_regress.c:27:15: warning: variable 'info' set but not used [-Wunused-but-set-variable]
   27 |     int i, j, info;
      |               ^~~~
[ 82%] Linking C executable openblas_utest
/proprietary/gcc/aarch64-linux-gnu/9.3.0/../../../../aarch64-linux-gnu/bin/ld: ../lib/libopenblas.a(zpotrs.f.o): Relocations in generic ELF (EM: 62)
/proprietary/gcc/aarch64-linux-gnu/9.3.0/../../../../aarch64-linux-gnu/bin/ld: ../lib/libopenblas.a(zpotrs.f.o): Relocations in generic ELF (EM: 62)
/proprietary/gcc/aarch64-linux-gnu/9.3.0/../../../../aarch64-linux-gnu/bin/ld: ../lib/libopenblas.a: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
make[2]: *** [utest/CMakeFiles/openblas_utest.dir/build.make:290: utest/openblas_utest] Error 1
make[1]: *** [CMakeFiles/Makefile2:636: utest/CMakeFiles/openblas_utest.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

**Comment by martin-frbg** [comment]
The error looks suspiciously like it is trying to include some objects that were mistakenly built for the x86_64 host earlier, and I wonder why it was picking up the system `ld` in the shared library build ? (It does not seem to do this in the static build, but was it using any other host tools (like ar, ranlib)

**Reply by Desperado17** [reply]
If you know how I can add more debug output just say it.

**Reply by martin-frbg** [reply]
sorry, `-DCMAKE_VERBOSE_MAKEFILE=ON` should do the trick (if you aren't using that already)

**Comment by Yaoxingtian** [comment]
Hi，did you solve this problem? could you share the way to solve it ? thanks !
